File: app/src/main/java/co/tpcreative/supersafe/common/adapter/BaseAdapter.java
package co.tpcreative.supersafe.common.adapter;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import co.tpcreative.supersafe.common.util.Utils;

public class BaseAdapter<V, VH extends BaseHolder> extends RecyclerView.Adapter<VH> {

    protected LayoutInflater inflater;
    protected List<V> dataSource = Collections.emptyList();
    private static final String TAG = BaseAdapter.class.getSimpleName();


    public BaseAdapter(LayoutInflater inflater) {
        this.inflater = inflater;
    }

    @Override
    public VH onCreateViewHolder(ViewGroup parent, int viewType) {
        return null;
    }

    @Override
    public void onBindViewHolder(VH holder, int position) {
        if (holder instanceof FooterViewHolder){
            return;
        }
        holder.bind(dataSource.get(position), position);
        holder.event();
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).hashCode();
    }

    public V getItem(int position){
        return dataSource.get(position);
    }

    @Override
    public int getItemCount() {
        return dataSource.size();
    }

    public void setDataSource(List<V> dataSource) {
      try {
        this.dataSource = new ArrayList<>(dataSource);
        notifyDataSetChanged();
      }catch (IllegalStateException e){      }
    }

    public List<V> getDataSource() {
        return dataSource;
    }

    public void appendItem(V item) {
        if (this.dataSource.isEmpty()) {
            this.dataSource = new ArrayList<>();
        }
        this.dataSource.add(item);
        notifyItemInserted(getItemCount());
    }

    public void removeAtPosition(int position){
        if(this.dataSource.size()> position){
            dataSource.remove(position);
            notifyItemRangeRemoved(position, 1);
        }
    }

    public void removeAt(int position) {
        try {
            Utils.Log(TAG,"Position "+ position +" size of dataSource "+dataSource.size());
            if(this.dataSource.size()> position){
                notifyItemRemoved(position);
                notifyItemRangeChanged(position,dataSource.size());
            }
            else{
                Utils.Log(TAG,"Can not delete ");
            }
        }
        catch (Exception e){
            e.printStackTrace();
        }
    }

    public void appendItems(@NonNull List<V> items) {
        if (dataSource.isEmpty()) {
            setDataSource(items);
        } else {
            int positionStart = getItemCount() - 1;
            dataSource.addAll(items);
            notifyItemRangeInserted(positionStart, items.size());
        }
    }

    public void addItemAtFirst(V item) {
        if (this.dataSource.isEmpty()) {
            this.dataSource = new ArrayList<>();
        }
        this.dataSource.add(0, item);
        notifyItemInserted(0);
    }

    public void addAtFirstAndRemoveEnd(V item) {
        if (this.dataSource.isEmpty()) {
            this.dataSource = new ArrayList<>();
        }
        this.dataSource.add(0, item);
        this.dataSource.remove(getItemCount() - 1);
        notifyItemRemoved(getItemCount() - 1);
        notifyItemInserted(0);
    }
}

File: app/src/main/java/co/tpcreative/supersafe/common/adapter/BaseAdapter.kt
package co.tpcreative.supersafe.common.adapter
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import co.tpcreative.supersafe.common.util.Utils

open class BaseAdapter<V, VH : BaseHolder<V>>(inflater: LayoutInflater) : RecyclerView.Adapter<VH>() {
    protected var inflater: LayoutInflater? = inflater
    protected var mDataSource: MutableList<V> = mutableListOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        TODO("Not yet implemented")
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        if (holder is FooterViewHolder<*>) {
            return
        }
        holder.bind(mDataSource[position]!!, position)
        holder.event()
    }

    override fun getItemId(position: Int): Long {
        return getItemId(position).hashCode().toLong()
    }

    fun getItem(position: Int): V? {
        return mDataSource[position]
    }

    override fun getItemCount(): Int {
        return mDataSource.size
    }

    fun setDataSource(dataSource: MutableList<V>?) {
        try {
            dataSource?.let {
                this.mDataSource = it
            } ?: run {
                this.mDataSource = mutableListOf()
            }
            notifyDataSetChanged()
        } catch (e: IllegalStateException) {
        }
    }

    fun getDataSource(): MutableList<V>? {
        return mDataSource
    }

    fun appendItem(item: V) {
        if (mDataSource.isEmpty()) {
            mDataSource = mutableListOf()
        }
        mDataSource.add(item)
        notifyItemInserted(itemCount)
    }

    fun removeAtPosition(position: Int) {
        if (mDataSource.size > position) {
            mDataSource.removeAt(position)
            notifyItemRangeRemoved(position, 1)
        }
    }

    fun removeAt(position: Int) {
        try {
            mDataSource.removeAt(position)
            notifyItemRemoved(position)
            notifyItemRangeChanged(position, mDataSource.size)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun appendItems(items: MutableList<V>) {
        if (mDataSource.isEmpty()) {
            setDataSource(items)
        } else {
            val positionStart = itemCount - 1
            mDataSource.addAll(items)
            notifyItemRangeInserted(positionStart, items.size)
        }
    }

    fun addItemAtFirst(item: V) {
        if (mDataSource.isEmpty()) {
            mDataSource = mutableListOf()
        }
        mDataSource.add(0, item)
        notifyItemInserted(0)
    }

    fun addAtFirstAndRemoveEnd(item: V) {
        if (mDataSource.isEmpty()) {
            mDataSource = mutableListOf()
        }
        mDataSource.add(0, item)
        notifyItemInserted(0)
        mDataSource.removeAt(itemCount - 1)
        notifyItemRemoved(itemCount - 1)
    }

    companion object {
        private val TAG = BaseAdapter::class.java.simpleName
    }

}